### PR TITLE
Feature/dispose delay

### PR DESCRIPTION
**Pull request by wjian23**

### CODE REVIEW

File: core/src/builtin_widgets/key.rs [hunk 1]
```diff
@@ -71,10 +72,11 @@ pub struct KeyWidget<V = ()> {
   #[declare(convert=into)]
   pub key: Key,
   pub value: Option<V>,
+
   #[declare(default)]
   before_value: Option<V>,
   #[declare(default)]
-  status: KeyStatus,
+  status: RefCell<KeyStatus>,
```
Comment (M-Adoo) [inline core/src/builtin_widgets/key.rs:79]: Should not wrap with `RefCell`, status is a public status, others may depend on it across methods like `is_enter`.

File: core/src/dynamic_widget.rs [hunk 2]
```diff
@@ -182,36 +204,113 @@ impl<D> DynRender<D> {
...
+      self.pin_widgets.borrow_mut().insert(wid, flag);
+      dirty_set.borrow_mut().insert(wid);
+    } else {
+      wid.remove_subtree(arena, store);
+    }
+  }
+
+  fn destroy_unhosted(&self, arena: &mut TreeArena, store: &mut LayoutStore) {
+    let mut widgets = self.pin_widgets.borrow_mut();
+    let to_remove = widgets
```
Comment (M-Adoo) [inline core/src/dynamic_widget.rs:304]: Needn't remove after collect, you can directly use `drain()` method.

File: core/src/builtin_widgets/pin_widget.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+use crate::{impl_query_self_only, prelude::*};
+use std::{cell::RefCell, rc::Rc};
+
+#[derive(Declare)]
+pub struct PinWidget {
+  #[declare(builtin, convert=custom)]
+  pub pin: Rc<RefCell<bool>>,
```
Comment (M-Adoo) [inline core/src/builtin_widgets/pin_widget.rs:7]: Public field should never be inner mutable.

File: core/src/dynamic_widget.rs (hunk 2 above)
Comment (wjian23) [reply core/src/dynamic_widget.rs]: drain_filter is unstable

File: widgets/src/themes/material/ripple.rs [hunk 4]
```diff
@@ -91,24 +98,23 @@ impl ComposeChild for Ripple {
...
-                let center = this.launch_pos.clone().unwrap();
-                Path::circle(center, radius, PathStyle::Fill)
+                Path::circle(launch_at, radius, PathStyle::Fill)
               }),
               from: Path::circle(Point::zero(), 0., PathStyle::Fill)
             }
+            Animate {
+              id: ripper_fade_out,
+              from: 0.,
+              prop: prop!(ripple.opacity, |from, to, rate| to.lerp(from,rate)),
```
Comment (M-Adoo) [inline widgets/src/themes/material/ripple.rs:108]: lerp_fn seems to same as default behavior.

File: core/src/dynamic_widget.rs [hunk 5]
```diff
@@ -80,11 +93,70 @@ impl<D: 'static> Render for DynRender<D> {
...
+    it: impl Iterator<Item = WidgetId>,
+    arena: &TreeArena,
+    set: &mut HashSet<WidgetId>,
+  ) {
+    it.for_each(|id| {
+      id.descendants(arena).for_each(|wid| {
+        wid
+          .assert_get(arena)
+          .query_on_first_type(QueryOrder::OutsideFirst, |w: &WidgetsHost| {
+            w.collect_subtree(w.wids.borrow().iter().cloned(), arena, set)
```
Comment (M-Adoo) [inline core/src/dynamic_widget.rs:148]: Seems unnecessary recursive collect.

File: core/src/builtin_widgets/delay_drop_widget.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  #[declare(builtin)]
+  pub delay_drop_until: bool,
+}
+
+impl ComposeChild for DelayDropWidget {
+  type Child = Widget;
+  #[inline]
+  fn compose_child(this: State<Self>, child: Self::Child) -> Widget {
+    let this = match this {
+      State::Stateless(w) => Stateful::new(w),
```
Comment (M-Adoo) [inline core/src/builtin_widgets/delay_drop_widget.rs:14]: `DelayDropWidget`  only used to query, why force it as stateful?

File: macros/src/builtin_fields_list.rs [hunk 7]
```diff
@@ -293,4 +293,9 @@ builtin! {
     #[doc= "return an observable stream of the widget disposed event"]
     disposed_stream: LifecycleSubject,
   }
+
+  DelayDropWidget {
+    #[doc= "the widget which delay_drop_until of true will still be hosted by it's parent event after disposing.it's useful to implement disappear animate "]
```
Comment (M-Adoo) [inline macros/src/builtin_fields_list.rs:298]: How about:

This widget delay the drop of its child until the field `delay_drop_until` is false, but not affect its `dispose` event emit time. It's useful to ensure the disappear-animate display fully.

File: core/src/dynamic_widget.rs (hunk 5 above)
Comment (wjian23) [reply core/src/dynamic_widget.rs:148]: necessary when a delay drop widget is hosted by another delay drop widget

**Review by M-Adoo — APPROVED** [review]

**Review by sologeek — APPROVED** [review]

**Pull request merged**